### DISCUSSION TITLE
Keep getting closing frames from Live Transcription WebSocket

### DISCUSSION THREAD

**Discussion by yunxianggu99** [deepgram/community#137] [General help]
Hi! I've been developing live transcription feature with Deepgram's [live streaming audio endpoint](https://developers.deepgram.com/api-reference/transcription/#transcribe-live-streaming-audio). Understand that it will get disconnected after eg I mute my microphone or switch audio input device. But when audio is recovered and I want to reconnect, I keep getting this closing frame(?):
 ```
 {:ok, %{"channels" => 1, "created" => "2023-05-04T04:06:56.048Z", "duration" => 15.239938, "model_info" => %{"3312b52b-d08e-4072-bfff-124c410b770d" => %{"arch" => "nova", "name" => "general-nova", "version" => "2023-03-13.31000"}}, "models" => ["3312b52b-d08e-4072-bfff-124c410b770d"], "request_id" => "6d752087-22e7-4e48-a170-8c24a9b88b81", "sha256" => "a31281a98cb68a674cdb163eb7bb87c71476d9205d89edaef7484b60f6352011", "transaction_key" => "deprecated"}}
```
This happenes even if I start a new socket and connect.

**Comment by nikolawhallon** [comment]
This looks like the final metadata websocket text message that Deepgram will send back when a stream is ending (immediately before sending a close websocket message) - I imagine this is then coming from the previous websocket connection you are closing, and not the new websocket connection you are creating.

**Reply by yunxianggu99** [reply]
But we are getting only this ending frame, even if we try to reconnect => no valid frames from Deepgram at all. Seems like all subsequent reconnections will be rejected

**Reply by nikolawhallon** [reply]
I have an idea - can you share the URL you are using (the full one with query parameters and whatnot), and also do you know the format of the audio you are sending (is it raw audio, or containerized)?

The response you posted shows that 15 seconds of audio was sent - for those 15 seconds, did you receive no text websocket messages from Deepgram?

**Reply by yunxianggu99** [reply]
So we are trying to do `wss://api.deepgram.com/v1/listen?punctuate=true&smart_format=true&tier=nova&language=en-US`
Think we have containerized audio from Twilio Video session (`audio/webm`)
And no, didn't receive text websocket messages

**Reply by yunxianggu99** [reply]
To give the whole story, we are trying to get remote audio tracks from Twilio Video session => then stream to deepgram for live transcription. But after remote track gets muted for ~1sec, will see this response

**Reply by nikolawhallon** [reply]
My original idea was - if using containerized audio, and the stream breaks/stops, and then you reconnect and continue streaming your audio, now to a new websocket connection with Deepgram, then the audio might be interpreted incorrectly, since header information from the beginning of the stream would not be present - but I'm fairly convinced that's not what is happening, especially with `audio/webm`

> But after remote track gets muted for ~1sec, will see this response

^ this might help understand - is there a chance that when a track gets muted you might at any point send an empty binary message to Deepgram? This happens to be a (I believe deprecated) way to tell Deepgram to "close the stream, and send me that last piece of summary metadata" - it's possible that for the duration your mic is muted, if somehow that causes you to send empty binary messages, that's why Deepgram will start immediately closing all connections you try to open.

**Reply by yunxianggu99** [reply]
Hmm that may well be the case. We try to remove tracks that are muted, but there could be delay. So is there a way to filter out binary messages that are empty? Or what is the suggested workaround?

**Reply by nikolawhallon** [reply]
For now, inspecting the size of the binary audio before each websocket send to Deepgram and skipping the send when the size is 0 is probably the best way - though this close feature is deprecated and might eventually no longer be around at some point in the future.

**Reply by yunxianggu99** [reply]
Thanks a lot!

**Reply by yunxianggu99** [reply]
Hey sorry but we are now making sure the blob size is > 0. 
But sometimes still see `<unk><unk><unk><unk>` in transcription result. Also still seeing the closing frame. Even if we try to reconnect and sending more audio blobs, no response is getting returned other than
```
{:ok, %{"channels" => 0, "created" => "2023-05-05T05:26:47.184Z", "duration" => 0.0, "request_id" => "3af31498-9378-4fe0-8fea-24c6b9f6e047", "sha256" => "43a83ea9f1575d25d63456ddc26e8c46f4cc62f741baa4e8108748676908c7ea", "transaction_key" => "deprecated"}}
```

**Reply by nikolawhallon** [reply]
Seeing `<unk><unk><unk><unk>` sounds like an issue on our end, I'll take a look, but I looked at that request id (`3af31498-9378-4fe0-8fea-24c6b9f6e047`) internally and see:
```
Failed to transcode: Client error: failed to process audio: corrupt or unsupported data
```
this will occur when not sending raw audio (which requires specifying the encoding, sample_rate, and channels) and when the audio is not properly containerized - when I run into these issues, I usually fork/dump the audio in my pipeline to a file to inspect to make sure the format is as expected. Let me look a bit more.

**Reply by nikolawhallon** [reply]
(Do you happen to have some docs which might describe Twilio Video's media format more? I'm trying to look into that - I know their programmable voice media format is strictly raw mulaw.)

**Reply by nikolawhallon** [reply]
There's a cool example of running a mock Deepgram server and validating your audio here which can be quite useful: https://developers.deepgram.com/documentation/guides/streaming-test-suite/#3-streaming-other-audio - it has utilities for dumping the streamed audio for inspection

**Reply by yunxianggu99** [reply]
So if we dump things into an audio for transcription, then everything works. Also the disconnection seems really random. Sometimes it comes back, sometimes transcription just keeps terminating

**Reply by nikolawhallon** [reply]
> So if we dump things into an audio for transcription, then everything works

do you mean if you take that audio to do a batch-mode request it will work?

If you are able to share some of the following, I can do some investigation:
* request id's where the disconnection occurs
* saved audio which worked or didn't work
* any code you are willing/able to share
